### PR TITLE
Expose the new parser using the --new-parser flag

### PR DESCRIPTION
**Pull request by certik**

### CODE REVIEW

**Comment by certik** [comment]
It's still WIP.

**Comment by Thirumalai-Shaktivel** [comment]
I'm working on this now, I will make the changes and push it here.

**Comment by Thirumalai-Shaktivel** [comment]
I'm working on this now, I will make the changes and push it here.

**Comment by Thirumalai-Shaktivel** [comment]
Test 1:
```console
$ lpython --show-ast --new-parser examples/expr2.py
syntax error: Token 'def' is unexpected here
 --> examples/expr2.py:1:1
  |
1 | def main0():
  | ^^^ 


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed).
```
Test 2: 
```console
$ lpython --show-ast --new-parser examples/expr2.py 
(Module [(Expr (BinOp (BinOp (BinOp (BinOp (BinOp (BinOp (Name a Load) Mult (Name z Load)) Add (ConstantInt 3 ())) Add (BinOp (ConstantInt 2 ()) Mult (Name x Load))) Add (BinOp (ConstantInt 3 ()) Mult (Name y Load))) Sub (BinOp (Name x Load) Div (BinOp (BinOp (Name z Load) Pow (ConstantInt 2 ())) Sub (ConstantInt 4 ())))) Sub (BinOp (Name x Load) Pow (BinOp (Name y Load) Pow (Name z Load)))))] [])
```

**Comment by Thirumalai-Shaktivel** [comment]
Any other changes to be made? 
cc @certik

File: src/bin/lpython.cpp [hunk 1]
```diff
@@ -138,18 +140,19 @@ int emit_asr(const std::string &infile,
...
     }
     LFortran::LPython::AST::ast_t* ast = r1.result;
 
     LFortran::LocationManager lm;
     lm.in_filename = infile;
-    std::string input = read_file(infile);
+    std::string input = LFortran::read_file(infile);
     lm.init_simple(input);
-    LFortran::diag::Diagnostics diagnostics;
+    diagnostics.diagnostics.clear();
```
Comment (certik) [inline src/bin/lpython.cpp:155]: Do we want to clear it? I think we want to accumulate all messages, such as warnings from the parser, wouldn't we?

File: src/lpython/parser/parser.cpp [hunk 2]
```diff
@@ -86,16 +86,18 @@ void Parser::handle_yyerror(const Location &loc, const std::string &msg)
...
     if (new_parser) {
         std::string input = read_file(infile);
-        Result<LPython::AST::Module_t*> res = parse(
-            al, input, diagnostics);
+        Result<LPython::AST::Module_t*> res = parse(al, input, diagnostics);
         LocationManager lm;
         lm.in_filename = infile;
         lm.init_simple(input);
+        CompilerOptions compiler_options;
+        std::cerr << diagnostics.render(input, lm, compiler_options);
```
Comment (certik) [inline src/lpython/parser/parser.cpp:100]: I thought about this, and I think we should not print the errors here, but rather return them to the caller, and the caller can decide when and how to visualize them. That way this parser function is fully general and can be used in all circumstances.

**Comment by certik** [comment]
Thanks!! Let me test it, and fix up a few things (see above) and merge it. This is super helpful.

File: src/bin/lpython.cpp (hunk 1 above)
Comment (certik) [reply src/bin/lpython.cpp:165]: I left this in, as currently we print the messages from the parser, so that we clear it, so that we don't print it again.

**Comment by certik** [comment]
I fixed it up, I think it works great now. We can now iteratively improve the parser.

File: src/lpython/parser/parser.cpp (hunk 2 above)
Comment (certik) [reply src/lpython/parser/parser.cpp]: I fixed it.

Comment (Thirumalai-Shaktivel) [reply src/lpython/parser/parser.cpp]: > I thought about this, and I think we should not print the errors here, but rather return them to the caller, and the caller can decide when and how to visualize them. That way this parser function is fully general and can be used in all circumstances.

Yup, I agree. Thanks for fixing this.

**Pull request merged**